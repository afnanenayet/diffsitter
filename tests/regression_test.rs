#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use libdiffsitter::{
        diff::compute_edit_script, generate_ast_vector_data, input_processing::TreeSitterProcessor,
        parse::GrammarConfig,
    };
    use std::path::PathBuf;
    use test_case::test_case;

    /// Get paths to input files for tests
    fn get_test_paths(test_type: &str, test_name: &str, ext: &str) -> (PathBuf, PathBuf) {
        let test_data_root = PathBuf::from(format!("./test_data/{test_type}/{test_name}"));
        let path_a = test_data_root.join(format!("a.{ext}"));
        let path_b = test_data_root.join(format!("b.{ext}"));
        assert!(
            path_a.exists(),
            "test data path {} does not exist",
            path_a.to_str().unwrap()
        );
        assert!(
            path_b.exists(),
            "test data path {} does not exist",
            path_b.to_str().unwrap()
        );

        (path_a, path_b)
    }

    #[test_case("short", "rust", "rs", true, true)]
    #[test_case("short", "python", "py", true, true)]
    #[test_case("short", "go", "go", true, true)]
    #[test_case("medium", "rust", "rs", true, false)]
    #[test_case("medium", "rust", "rs", false, false)]
    #[test_case("medium", "cpp", "cpp", true, true)]
    #[test_case("medium", "cpp", "cpp", false, true)]
    fn diff_hunks_snapshot(
        test_type: &str,
        name: &str,
        ext: &str,
        split_graphemes: bool,
        strip_whitespace: bool,
    ) {
        let (path_a, path_b) = get_test_paths(test_type, name, ext);
        let config = GrammarConfig::default();
        let ast_data_a = generate_ast_vector_data(path_a, None, &config).unwrap();
        let ast_data_b = generate_ast_vector_data(path_b, None, &config).unwrap();

        let processor = TreeSitterProcessor {
            split_graphemes,
            strip_whitespace,
            ..Default::default()
        };

        let diff_vec_a = processor.process(&ast_data_a.tree, &ast_data_a.text);
        let diff_vec_b = processor.process(&ast_data_b.tree, &ast_data_b.text);
        let diff_hunks = compute_edit_script(&diff_vec_a, &diff_vec_b).unwrap();

        // We have to set the snapshot name manually, otherwise there appear to be threading issues
        // and we end up with more snapshot files than there are tests, which cause
        // nondeterministic errors.
        let snapshot_name = format!("{test_type}_{name}_split_graphemes_{split_graphemes}_strip_whitespace_{strip_whitespace}");
        assert_debug_snapshot!(snapshot_name, diff_hunks);
    }
}
