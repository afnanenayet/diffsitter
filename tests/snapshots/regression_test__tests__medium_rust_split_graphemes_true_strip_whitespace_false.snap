---
source: tests/regression_test.rs
expression: snapshot_string
---
New(Line={line_index=2, entries=
[
    Entry{',', start=(2, 26), end=(2, 27)}
]}
)
New(Line={line_index=20, entries=
[
    Entry{'_', start=(20, 12), end=(20, 13)}
    Entry{'f', start=(20, 13), end=(20, 14)}
    Entry{'n', start=(20, 14), end=(20, 15)}
]}
)
New(Line={line_index=42, entries=
[
    Entry{',', start=(42, 19), end=(42, 20)}
]}
)
New(Line={line_index=59, entries=
[
    Entry{'b', start=(59, 4), end=(59, 5)}
    Entry{'l', start=(59, 5), end=(59, 6)}
    Entry{'e', start=(59, 6), end=(59, 7)}
    Entry{'a', start=(59, 7), end=(59, 8)}
]}
)
Old(Line={line_index=53, entries=
[
    Entry{'a', start=(53, 8), end=(53, 9)}
    Entry{'l', start=(53, 9), end=(53, 10)}
    Entry{'k', start=(53, 10), end=(53, 11)}
]}
)
Old(Line={line_index=55, entries=
[
    Entry{'p', start=(55, 21), end=(55, 22)}
    Entry{'u', start=(55, 23), end=(55, 24)}
    Entry{'s', start=(55, 24), end=(55, 25)}
    Entry{'e', start=(55, 25), end=(55, 26)}
]}
)
New(Line={line_index=61, entries=
[
    Entry{'b', start=(61, 15), end=(61, 16)}
    Entry{'e', start=(61, 16), end=(61, 17)}
    Entry{'t', start=(61, 18), end=(61, 19)}
]}
)
New(Line={line_index=67, entries=
[
    Entry{'e', start=(67, 9), end=(67, 10)}
    Entry{'d', start=(67, 10), end=(67, 11)}
    Entry{'L', start=(67, 34), end=(67, 35)}
    Entry{'g', start=(67, 36), end=(67, 37)}
    Entry{'a', start=(67, 37), end=(67, 38)}
    Entry{'n', start=(67, 38), end=(67, 39)}
]}
)
Old(Line={line_index=61, entries=
[
    Entry{'d', start=(61, 12), end=(61, 13)}
    Entry{'o', start=(61, 13), end=(61, 14)}
    Entry{'l', start=(61, 14), end=(61, 15)}
    Entry{'l', start=(61, 15), end=(61, 16)}
    Entry{'y', start=(61, 16), end=(61, 17)}
    Entry{'D', start=(61, 40), end=(61, 41)}
    Entry{'l', start=(61, 42), end=(61, 43)}
    Entry{'l', start=(61, 43), end=(61, 44)}
    Entry{'y', start=(61, 44), end=(61, 45)}
]}
)
New(Line={line_index=70, entries=
[
    Entry{'e', start=(70, 1), end=(70, 2)}
    Entry{'.', start=(70, 3), end=(70, 4)}
    Entry{'b', start=(70, 4), end=(70, 5)}
    Entry{'l', start=(70, 5), end=(70, 6)}
    Entry{'e', start=(70, 6), end=(70, 7)}
    Entry{'a', start=(70, 7), end=(70, 8)}
]}

Line={line_index=71, entries=
[
    Entry{'e', start=(71, 1), end=(71, 2)}
    Entry{'d', start=(71, 2), end=(71, 3)}
]}

Line={line_index=72, entries=
[
    Entry{'e', start=(72, 1), end=(72, 2)}
    Entry{'.', start=(72, 3), end=(72, 4)}
    Entry{'b', start=(72, 4), end=(72, 5)}
    Entry{'l', start=(72, 5), end=(72, 6)}
    Entry{'e', start=(72, 6), end=(72, 7)}
    Entry{'a', start=(72, 7), end=(72, 8)}
]}
)
Old(Line={line_index=64, entries=
[
    Entry{'o', start=(64, 5), end=(64, 6)}
    Entry{'l', start=(64, 6), end=(64, 7)}
    Entry{'l', start=(64, 7), end=(64, 8)}
    Entry{'y', start=(64, 8), end=(64, 9)}
    Entry{'.', start=(64, 9), end=(64, 10)}
    Entry{'a', start=(64, 11), end=(64, 12)}
    Entry{'l', start=(64, 12), end=(64, 13)}
    Entry{'k', start=(64, 13), end=(64, 14)}
]}

Line={line_index=65, entries=
[
    Entry{'d', start=(65, 4), end=(65, 5)}
    Entry{'o', start=(65, 5), end=(65, 6)}
    Entry{'l', start=(65, 6), end=(65, 7)}
    Entry{'l', start=(65, 7), end=(65, 8)}
    Entry{'y', start=(65, 8), end=(65, 9)}
]}

Line={line_index=66, entries=
[
    Entry{'o', start=(66, 5), end=(66, 6)}
    Entry{'l', start=(66, 6), end=(66, 7)}
    Entry{'l', start=(66, 7), end=(66, 8)}
    Entry{'y', start=(66, 8), end=(66, 9)}
    Entry{'.', start=(66, 9), end=(66, 10)}
    Entry{'a', start=(66, 11), end=(66, 12)}
    Entry{'l', start=(66, 12), end=(66, 13)}
    Entry{'k', start=(66, 13), end=(66, 14)}
]}
)
